File: tree/297. Serialize and Deserialize Binary Tree/py/main.py
from collections import deque

class TreeNode(object):
  def __init__(self, x):
    self.val = x
    self.left = None
    self.right = None

class Codec:
    """Encodes a tree to a single string.    
      :type root: TreeNode
      :rtype: str
    """
    def serialize(self, root):
        if root == None:
          return []

        queue = deque()
        queue.append(root)
        res = []

        while queue:
          node = queue.popleft()

          if node:
            queue.append(node.left)
            queue.append(node.right)
            res.append(node.val)
          else:
            res.append(None)

        while len(res) and res[-1] == None:
          res.pop()

        return res
        
    """
      Decodes your encoded data to tree.  
      :type data: str
      :rtype: TreeNode
    """
    def deserialize(self, data):
      if len(data) == 0:
        return None

      queue = deque()
      index = 0
      
      root = TreeNode(data[index])
      index += 1

      queue.append(root)

      while queue and index < len(data):
        node = queue.popleft()
        left = data[index]
        index += 1

        if left != None:
          node.left = TreeNode(left)
          queue.append(node.left)
            
        if index == len(data):
            break
            
        right = data[index]
        index += 1
        
        if right != None:
          node.right = TreeNode(right)
          queue.append(node.right)

      return root        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: tree/297. Serialize and Deserialize Binary Tree/js/main.js
class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
const serialize = root => {
  if (root == null) return [];

  let nums = [];
  let queue = [];

  queue.push(root);

  while (queue.length) {
    let node = queue.shift();

    if (node) {
      nums.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      nums.push(null);
    }
  }

  while (nums.length && nums[nums.length - 1] == null) {
    nums.pop();
  }

  return nums;
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {[]} data
 * @return {TreeNode}
 */
const deserialize = data => {
  if (data.length == 0) return null;

  let queue = [];
  let index = 0;

  let root = new TreeNode(data[index++]);
  queue.push(root);

  while (queue.length) {
    let node = queue.shift();

    let left = data[index++];

    if (left != null && left != undefined) {
      node.left = new TreeNode(left);
      queue.push(node.left);
    }

    let right = data[index++];

    if (right != null && right != undefined) {
      node.right = new TreeNode(right);
      queue.push(node.right);
    }
  }

  return root;
};
